### ISSUE TITLE
Method resolution fail

### ISSUE THREAD

**Issue by rnesytov** [antonmedv/expr#353]
The VM tries to resolve a method by its index, but sometimes it fails.
Example:
```go
type (
	Ctx interface {
		GetValue(string) Value
	}
	Value interface {
		Number() int
	}

	ctx struct{}
	val struct{}
)

func (c *ctx) GetValue(key string) Value {
	return val{}
}

func (v val) Number() int {
	return 1
}

func (v val) MarshallJSON() ([]byte, error) {
	return []byte{}, nil
}

func NewCtx() Ctx {
	return &ctx{}
}

func TestMethodsError(t *testing.T) {
	env := map[string]any{
		"Ctx": NewCtx(),
	}
	// with expr.Eval
	_, err := expr.Eval("Ctx.GetValue('a').Number()", env)
	if err != nil {
		t.Fatal(err)
	}
	// with expr.Run
	program, err := expr.Compile("Ctx.GetValue('a').Number()", expr.Env(env))
	if err != nil {
		t.Fatal(err)
	}
	output, err := expr.Run(program, env)

	if err != nil {
		t.Fatal(err)
	}
	t.Log(output)
}
```
The test successfully passes the first step (with expr.Eval), but fails on the second step (with expr.Compile + expr.Run) with the following error message:
```
interface conversion: interface {} is func() ([]uint8, error), not func() int (1:19)
         | Ctx.GetValue('a').Number()
         | ..................^
```

As I see the problem is here:
https://github.com/antonmedv/expr/blob/e10a35d3dffdbd6ec83ca65b2f76d6a12ab373b8/vm/runtime/runtime.go#L134

The method is fetched by its index in the interface, but in this case, `v` refers to a struct that implements the interface and has a different index for the method.

It looks like it could be fixed by replacing the `v.Method(method.Index)` call with `v.MethodByName(method.Name)`.

**Comment by antonmedv** [comment]
Can you show generated opcode? program.Disassemble()

**Comment by rnesytov** [comment]
@antonmedv 
```
0	OpPush	0	a
1	OpLoadFast	1	Ctx
2	OpMethod	2	{GetValue 0}
3	OpCall	1
4	OpMethod	3	{Number 0}
5	OpCallTyped	10
```

**Comment by antonmedv** [comment]
Looks like for some reason `MarshallJSON` method was selected instead of `Number()`. I'll try to debug it.

**Comment by antonmedv** [comment]
Fixed in 1a1138d22ab8eca39eb0286dd63299d0ce57a4c6 and added the test for this use case.

Released [v1.12.4](https://github.com/antonmedv/expr/releases/tag/v1.12.4).
